include/solvers/PreconditionCG.hpp

```diff
@@ -0,0 +1,167 @@
+// Copyright (c)   2017 Tyler Olsen
+//                 2018 Patrick Diehl
+//                 2019 Nanmiao Wu
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+
+#ifndef BLAZE_ITERATIVE_PRECONDITIONCG_HPP
+#define BLAZE_ITERATIVE_PRECONDITIONCG_HPP
+
+#include "PreconditionCGTag.hpp"
+
+BLAZE_NAMESPACE_OPEN
+ITERATIVE_NAMESPACE_OPEN
+
+    namespace detail {
+        // Decompose A as A = L + D + U, where L is strictly lower triangular matrix, D is diagonal matrix, and U is
+        // strictly upper triangular matrix.
+        template<typename MatrixType, typename T>
+        void preconditioner_matrix(std::string type, const MatrixType &A, MatrixType &M){
+
+            if (type.compare("Jacobi preconditioning")==0){
+
+                DiagonalMatrix<MatrixType> D(A.columns());
+                band<0L>(D) = band<0L>(A);
+                M = D;
+            }
+
+            if (type.compare("Gauss Seidel preconditioning")==0){
+                DiagonalMatrix<MatrixType> D(A.columns());
+                StrictlyLowerMatrix<MatrixType> L(A.columns());
+
+                for(std::size_t i = 1 - A.columns(); i < 0; ++i){
```

Comment:
`std::size_t` is an unsigned type, while this apparently works I'd suggest to use some signed integral type as the loop variable. Same is true for other loops in this PR as well.